### PR TITLE
Add indexing enhancements and version to add-on docs

### PR DESCRIPTION
**Pull request by vinistock**
### Motivation

This PR documents indexing enhancements and adds the `version` requirement to our add-on docs.

### CODE REVIEW

File: jekyll/add-ons.markdown [hunk 1]
```diff
@@ -126,6 +132,45 @@ This approach enables all add-on responses to be captured in a single round of A
 
 ### Enhancing features
 
+There are two ways to enhance Ruby LSP features. One is handling DSLs that occur at a call site and that do not change
+which declarations exist in the project. A great example of this the Rails validation method, which accepts a symbol
```
Comment (andyw8) [inline jekyll/add-ons.markdown:136]:
```suggestion
which declarations exist in the project. A great example of this the Rails `validate` method, which accepts a symbol
```

File: jekyll/add-ons.markdown [hunk 2]
```diff
@@ -126,6 +132,45 @@ This approach enables all add-on responses to be captured in a single round of A
...
+that represents a method that gets dynamically invoked. That style of DSL is what we refer to as a [call site
+DSL](#dealing-with-call-site-dsls).
```
Comment (andyw8) [inline jekyll/add-ons.markdown:138]: Would it be more correct to call this a 'DSL method' or something like that?

File: jekyll/add-ons.markdown [hunk 3]
```diff
@@ -195,6 +244,103 @@ module RubyLsp
 end
 ```
 
+#### Dealing with declaration DSLs
+
+Add-ons can inform the Ruby LSP about declarations that are made via meta-programming. By ensuring that the index is
+populated with all declarations, features like go to definition, hover, completion, signature help and workspace symbol
+will all automatically work.
+
+To achieve this the add-on must create an indexing enhancement class and then it needs to register that. Here's an
```
Comment (andyw8) [inline jekyll/add-ons.markdown:253]:
```suggestion
To achieve this the add-on must create an indexing enhancement class and register it. Here's an
```

File: jekyll/add-ons.markdown [hunk 4]
```diff
@@ -195,6 +244,103 @@ module RubyLsp
...
+    # Create the array of signatures that this method will accept. Every signatures is composed of a list of
+    # parameters. The parameter classes represent each type of parameter
+    signatures = [
+      Entry::Signature.new([Entry::RequiredParameter.new(name: :a)])
+    ]
+
+    new_entry = Entry::Method.new(
+      "new_method", # The name of the method that gets created via meta-programming
+      file_path,    # The file_path where the DSL call was found. This should always just be the file_path received
+      location,     # The location where the DSL call was found
```
Comment (andyw8) [inline jekyll/add-ons.markdown:300]:
```suggestion
      location,     # The Prism node location where the DSL call was found
```

File: jekyll/add-ons.markdown [hunk 5]
```diff
@@ -346,14 +504,56 @@ end
...
+For add-ons that have a runtime dependency on the `ruby-lsp` gem, you can simply use regular gemspec constraints to
+define which version is supported.
 
 ```ruby
 spec.add_dependency("ruby-lsp", "~> 0.6.0")
 ```
 
+#### Add-ons that do not have a runtime dependency on the ruby-lsp
+
+For add-ons that are defined inside other gems that do not wish to take a runtime dependency on `ruby-lsp`, please use
```
Comment (andyw8) [inline jekyll/add-ons.markdown:522]:
```suggestion
For add-ons that are defined inside other gems that do not wish to have a runtime dependency on `ruby-lsp`, please use
```

**Comment by andyw8** [comment]
@alexcrocha it would be good to get your input here as a 'newcomer' to developing add-ons.

File: jekyll/add-ons.markdown (hunk 2 above)
Comment (vinistock) [reply jekyll/add-ons.markdown:138]: I'm not sure what the best term is. A method that mutates the index like `belongs_to` is also a "DSL method".

I was looking for terms that can differentiate DSL on the call site vs DSLs that create declarations.

Comment (andyw8) [reply jekyll/add-ons.markdown:138]: I was meaning just the 'DSL' part, to me the 'DSL' is the _collection_ of all the available 'constructs'. Not a big deal though.

**Pull request merged**